packages/cw721/src/traits.rs

```diff
@@ -0,0 +1,791 @@
+        Ok(())
+    }
+}
+
+pub trait ToAttributesState {
+    fn to_attributes_state(&self) -> Result<Vec<Attribute>, Cw721ContractError>;
+}
+
+impl<T> ToAttributesState for Option<T>
+where
+    T: ToAttributesState,
+{
+    fn to_attributes_state(&self) -> Result<Vec<Attribute>, Cw721ContractError> {
+        match self {
+            Some(inner) => inner.to_attributes_state(),
+            None => Ok(vec![]),
+        }
+    }
+}
+
+pub trait FromAttributesState: Sized {
+    fn from_attributes_state(value: &[Attribute]) -> Result<Self, Cw721ContractError>;
+}
+
+impl<T> FromAttributesState for Option<T>
+where
+    T: FromAttributesState,
+{
+    fn from_attributes_state(value: &[Attribute]) -> Result<Self, Cw721ContractError> {
+        if value.is_empty() {
+            Ok(None)
+        } else {
+            T::from_attributes_state(value).map(Some)
+        }
+    }
+}
+
+pub trait Cw721Execute<
+    // NftInfo extension (onchain metadata).
+    TNftExtension,
+    // NftInfo extension msg for onchain metadata.
+    TNftExtensionMsg,
+    // CollectionInfo extension (onchain attributes).
+    TCollectionExtension,
+    // CollectionInfo extension msg for onchain collection attributes.
+    TCollectionExtensionMsg,
+    // Custom extension msg for custom contract logic. Default implementation is a no-op.
+    TExtensionMsg,
+    // Defines for `CosmosMsg::Custom<T>` in response. Barely used, so `Empty` can be used.
+    TCustomResponseMsg,
+> where
+    TNftExtension: Cw721State,
+    TNftExtensionMsg: Cw721CustomMsg + StateFactory<TNftExtension>,
+    TCollectionExtension: Cw721State + ToAttributesState + FromAttributesState,
+    TCollectionExtensionMsg: Cw721CustomMsg + StateFactory<TCollectionExtension>,
+    TCustomResponseMsg: CustomMsg,
+{
+    fn instantiate_with_version(
+        &self,
+        deps: DepsMut,
+        env: &Env,
+        info: &MessageInfo,
+        msg: Cw721InstantiateMsg<TCollectionExtensionMsg>,
+        contract_name: &str,
+        contract_version: &str,
+    ) -> Result<Response<TCustomResponseMsg>, Cw721ContractError> {
+        instantiate_with_version(deps, env, info, msg, contract_name, contract_version)
+    }
+
+    fn instantiate(
+        &self,
+        deps: DepsMut,
+        env: &Env,
+        info: &MessageInfo,
+        msg: Cw721InstantiateMsg<TCollectionExtensionMsg>,
+    ) -> Result<Response<TCustomResponseMsg>, Cw721ContractError> {
+        instantiate(deps, env, info, msg)
+    }
+
+    fn execute(
+        &self,
+        deps: DepsMut,
+        env: &Env,
+        info: &MessageInfo,
+        msg: Cw721ExecuteMsg<TNftExtensionMsg, TCollectionExtensionMsg, TExtensionMsg>,
+    ) -> Result<Response<TCustomResponseMsg>, Cw721ContractError> {
+        match msg {
+            Cw721ExecuteMsg::UpdateCollectionInfo { collection_info } => {
+                self.update_collection_info(deps, info.into(), env.into(), collection_info)
+            }
```

Comment:
`Cw721Execute` trait with default implementation. Like the entry point for updating collection info.